server/core/include/process_manager.hpp

```diff
@@ -0,0 +1,185 @@
+#ifndef IRODS_PROCESS_MANAGER_HPP
+#define IRODS_PROCESS_MANAGER_HPP
+
+#include <unistd.h>
+#include <functional>
+#include <chrono>
+#include <any>
+#include <thread>
+#include <boost/lexical_cast.hpp>
+#include <memory>
+
+#include "irods_logger.hpp"
+
+/// \file
+/// \parblock
+/// The main use case of this facility is to provide a means of running tasks intermittently.
+/// that is, in an interval.
+/// \endparblock
+/// TODO add since information
+
+namespace irods::experimental::cron{
+    namespace detail{
+        /// \brief Placeholder function for cron_task
+        void noop(std::any&) noexcept {}
+    } // namespace detail
+
+    /// \brief a cron task, that is, one run intermittently.
+    ///
+    /// TODO add some /since information once it's known what release it's going to be in.
+    class cron_task{
+        using func = std::function<void(std::any&)>;
+        /// \brief the main function in the task
+        std::function<void(std::any&)> execute = detail::noop;
+        /// \brief The function to be run before the main function of the task
+        /// run only once when #run_pre_once_only is true
+        std::function<void(std::any&)> pre = detail::noop;
+        /// \brief The function to be run after the main function of the task
+        /// run only once when #run_post_once_only is true
+        std::function<void(std::any&)> post = detail::noop;
+        /// \brief The length of time between runs, in seconds
+        std::chrono::seconds interval = std::chrono::seconds(0);
+        /// \brief The timestamp of the last time this task was run
+        std::chrono::time_point<std::chrono::system_clock> last_run =
+            std::chrono::time_point<std::chrono::system_clock>(std::chrono::seconds(0));
+        /// \brief Only run the pre function once.
+        bool run_pre_once_only = true;
+        /// \brief Run post only once.
+        bool run_post_once_only  = true;
+        /// \brief User defined data to be handled by the task.
+        std::any data;
+        cron_task(func e, func pre, func post, std::chrono::seconds i, bool run_pre_once, bool run_post_once,
+                  std::any data):
+            execute(e),
+            pre(pre),
+            post(post),
+            interval(i),
+            run_pre_once_only(run_pre_once),
+            run_post_once_only(run_post_once),
+            data(data)
+        {
+        }
+    public:
+
+        ///  \brief Returns true if the task is ready to run
+        /// \retval true if sufficient time has elapsed
+        bool ready() const noexcept {
+            auto now = std::chrono::system_clock::now();
+            return (now - last_run) > interval;
+        }
+        /**
+         * \brief Returns true if the task has run before, as determined by the last run timestamp
+         *
+         * \retval true if the task has run before
+         */
+        bool has_run()const noexcept{
+            return last_run.time_since_epoch().count() == 0;
+        }
```

Comment:
is this logic correct?

following a successful run, this will return false.

Reply:
Yeah, that is definitely a mistake